src/clients/arclib/arctest.1.in
.TH arctest 1 "@DATE@" "NorduGrid ARC @VERSION@" "NorduGrid ARC"

.SH NAME

arctest \- ARC Test Suite

.SH DESCRIPTION

The
.B arctest
command tests basic client and server setup.

.SH SYNOPSIS

.B arctest [options]

.SH OPTIONS

.IP "\fB-c\fR, \fB--cluster\fR=\fIname\fR"
select one or more computing elements: \fIname\fR can be an alias for a single CE, a group of CEs or a URL
.IP "\fB-g\fR, \fB--index\fR=\fIname\fR"
select one or more registries: \fIname\fR can be an alias for a single registry, a group of registries or a URL
.IP "\fB-R\fR, \fB--rejectdiscovery\fR=\fIURL\fR"
skip the service with the given URL during service discovery
.IP "\fB-S\fR, \fB--submissioninterface\fR=\fIInterfaceName\fR"
only use this interface for submitting (e.g. org.nordugrid.gridftpjob, org.ogf.glue.emies.activitycreation, org.ogf.bes)
.IP "\fB-J\fR \fIjobid\fR, \fB--jobid\fR=\fIjobid\fR"
submits testjob given by the jobid.
.IP "\fB-r\fR \fItime\fR, \fB--runtime\fR=\fItime\fR"
test job runtime specified in case of the 1st test job.
.IP "\fB-j\fR, \fB--joblist\fR=\fIfilename\fR"
the file storing information about active jobs (default ~/.arc/jobs.xml)
.IP "\fB-o\fR, \fB--jobids-to-file\fR=\fIfilename\fR"
the IDs of the submitted jobs will be appended to this file
.IP "\fB-D\fR, \fB--dryrun\fR"
submit jobs as dry run (no submission to batch system)
.IP "\fB-x\fR, \fB--dumpdescription\fR"
do not submit - dump job description in the language accepted by the target
.IP "\fB-E\fR, \fB--certificate\fR"
prints information about installed user- and CA-certificates
.IP "\fB-P\fR, \fB--listplugins\fR"
list the available plugins
.IP "\fB-t\fR, \fB--timeout\fR=\fIseconds\fR"
timeout in seconds (default 20)
.IP "\fB-z\fR, \fB--conffile\fR=\fIfilename\fR"
configuration file (default ~/.arc/client.conf)
.IP "\fB-d\fR, \fB--debug\fR=\fIdebuglevel\fR"
FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG
.IP "\fB-b\fR, \fB--broker\fR=\fIbroker\fR"
selected broker: Random (default), FastestQueue or custom
.IP "\fB-v\fR, \fB-version\fR"
print version information
.IP "\fB-h\fR, \fB-help\fR"
print usage information
.LP

.SH EXTENDED DESCRIPTION

The
.B arctest
command is used for basic testing of client and server installation
through submitting various test jobs. It can also print basic information
about user's certificate. The command is complementary to
.B arcinfo
and
.B arcproxy -I

The test jobs available in this version of arctest are:

Test job 1: This test-job calculates prime-numbers for a number of minutes given by
.B -r
(default 5) and outputs the list to stderr. The source-code for the prime-number
program, the Makefile and the executable are downloaded to the cluster from HTTP
and FTP servers and the program is compiled before running. In this way, the test
job constitutes a fairly comprehensive test of the basic setup of a grid cluster.

Test job 2: attempts to list all environment variables at the remote site

Test job 3: copies a remote file from an HTTP server into a local file

.SH EXAMPLES

arctest -J 1 -c <clustername>

will submit test job number 1 to the specified cluster belonging to the flavor of it.

arctest --certificate

will print basic information about the user's certificate.

.SH AUTHOR

Ivan Marton <martoni@niif.hu>

.SH SEE ALSO

.BR arccat (1),
.BR arcclean (1),
.BR arccp (1),
.BR arcget (1),
.BR arcinfo (1),
.BR arckill (1),
.BR arcls (1),
.BR arcmigrate (1),
.BR arcmkdir (1),
.BR arcproxy (1),
.BR arcrenew (1),
.BR arcresub (1),
.BR arcresume (1),
.BR arcrm (1),
.BR arcstat (1),
.BR arcsub (1),
.BR arcsync (1)



NEWS
no news good news :)


src/clients/arclib/arctest.cpp
// -*- indent-tabs-mode: nil -*-

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <fstream>
#include <iostream>
#include <list>
#include <string>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#include <arc/ArcConfig.h>
#include <arc/ArcLocation.h>
#include <arc/DateTime.h>
#include <arc/IString.h>
#include <arc/Logger.h>
#include <arc/StringConv.h>
#include <arc/URL.h>
#include <arc/Utils.h>
#include <arc/compute/ComputingServiceRetriever.h>
#include <arc/compute/Job.h>
#include <arc/compute/SubmitterPlugin.h>
#include <arc/compute/JobDescription.h>
#include <arc/credential/Credential.h>
#include <arc/UserConfig.h>
#include <arc/compute/Broker.h>

#include "utils.h"

static Arc::Logger logger(Arc::Logger::getRootLogger(), "arcsub");

int test(const Arc::UserConfig& usercfg, Arc::ExecutionTargetSorter& ets, const Arc::JobDescription& testJob, const std::string& jobidfile);
int dumpjobdescription(const Arc::UserConfig& usercfg, Arc::ExecutionTargetSorter& ets, const Arc::JobDescription& testJob);

int RUNMAIN(arctest)(int argc, char **argv) {

  setlocale(LC_ALL, "");

  Arc::LogStream logcerr(std::cerr);
  logcerr.setFormat(Arc::ShortFormat);
  Arc::Logger::getRootLogger().addDestination(logcerr);
  Arc::Logger::getRootLogger().setThreshold(Arc::WARNING);

  Arc::ArcLocation::Init(argv[0]);

  ClientOptions opt(ClientOptions::CO_TEST,
                    istring(" "),
                    istring("The arctest command is used for "
                            "testing clusters as resources."));

  std::list<std::string> params = opt.Parse(argc, argv);

  if (opt.showversion) {
    std::cout << Arc::IString("%s version %s", "arctest", VERSION)
              << std::endl;
    return 0;
  }

  if ((opt.testjobid == -1) && (!opt.show_credentials) && (!opt.show_plugins)) {
    std::cout << Arc::IString("Nothing to do:\n"
        "you have to either specify a test job id with -J (--job)\n"
        "or query information about the certificates with -E (--certificate)\n");
    return 0;
  }

  if ((opt.testjobid == 1) && (!opt.runtime)) {
    std::cout << Arc::IString("For the 1st test job"
        "you also have to specify a runtime value with -r (--runtime) option.");
    return 0;
  }

  // If debug is specified as argument, it should be set before loading the configuration.
  if (!opt.debug.empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(opt.debug));


  Arc::UserConfig usercfg(opt.conffile, opt.joblist);
  if (!usercfg) {
    logger.msg(Arc::ERROR, "Failed configuration initialization");
    return 1;
  }

  if (opt.show_plugins) {
    std::list<std::string> types;
    types.push_back("HED:SubmitterPlugin");
    types.push_back("HED:ServiceEndpointRetrieverPlugin");
    types.push_back("HED:TargetInformationRetrieverPlugin");
    types.push_back("HED:JobDescriptionParser");
    types.push_back("HED:BrokerPlugin");
    showplugins("arctest", types, logger, usercfg.Broker().first);
    return 0;
  }

  if (!checkproxy(usercfg)) {
    return 1;
  }

  if (opt.debug.empty() && !usercfg.Verbosity().empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(usercfg.Verbosity()));

  if (opt.timeout > 0)
    usercfg.Timeout(opt.timeout);

  if (!opt.broker.empty())
    usercfg.Broker(opt.broker);

  if (opt.show_credentials) {
    std::string proxy_path = usercfg.ProxyPath();
    std::string cert_path = usercfg.CertificatePath();
    std::string key_path = usercfg.KeyPath();
    std::string ca_dir = usercfg.CACertificatesDirectory();

    const Arc::Time now;

    std::cout << Arc::IString("Certificate information:") << std::endl << std::endl;

    if (cert_path.empty() || key_path.empty()) {
      std::cout << Arc::IString("No user-certificate found") << std::endl << std::endl;
    } else {
      Arc::Credential holder(cert_path, "", ca_dir, "");
      std::cout << Arc::IString("Certificate: %s", cert_path) << std::endl;
      std::cout << Arc::IString("Subject name: %s", holder.GetDN()) << std::endl;
      std::cout << Arc::IString("Valid until: %s", (std::string) holder.GetEndTime() ) << std::endl << std::endl;
    }

    if (proxy_path.empty()) {
      std::cout << Arc::IString("No proxy found") << std::endl << std::endl;
    } else {
      Arc::Credential holder(proxy_path, "", ca_dir, "");
      std::cout << Arc::IString("Proxy: %s", proxy_path) << std::endl;
      std::cout << Arc::IString("Proxy-subject: %s", holder.GetDN()) << std::endl;
      if (holder.GetEndTime() < now) {
        std::cout << Arc::IString("Valid for: Proxy expired") << std::endl << std::endl;
      } else if (!holder.GetVerification()) {
        std::cout << Arc::IString("Valid for: Proxy not valid") << std::endl << std::endl;
      } else {
        std::cout << Arc::IString("Valid for: %s", (holder.GetEndTime() - now).istr()) << std::endl << std::endl;
      }
    }

    if (!cert_path.empty() && !key_path.empty()) {
      Arc::Credential holder(cert_path, "", ca_dir, "");
      std::cout << Arc::IString("Certificate issuer: %s", holder.GetIssuerName()) << std::endl << std::endl; //TODO
    }

    return EXIT_SUCCESS;
  }

  Arc::JobDescription testJob;
  if (!Arc::JobDescription::GetTestJob(opt.testjobid, testJob)) {
    std::cout << Arc::IString("No test-job, with ID \"%d\"", opt.testjobid) << std::endl;
    return 1;
  }

  // Set user input variables into job description
  if (opt.testjobid == 1) {
    for ( std::map<std::string, std::string>::iterator iter = testJob.OtherAttributes.begin();
      iter != testJob.OtherAttributes.end(); ++iter ) {
        char buffer [iter->second.length()+255];
        sprintf(buffer, iter->second.c_str(), opt.runtime+1, opt.runtime+3);
        iter->second = (std::string) buffer;
      }
  }

  Arc::Broker broker(usercfg, testJob, usercfg.Broker().first);
  if (!broker.isValid()) {
    logger.msg(Arc::ERROR, "Unable to load broker %s", usercfg.Broker().first);
    return 1;
  }
  logger.msg(Arc::INFO, "Broker %s loaded", usercfg.Broker().first);

  std::list<Arc::Endpoint> services = getServicesFromUserConfigAndCommandLine(usercfg, opt.indexurls, opt.clusters, opt.requestedSubmissionInterfaceName, opt.infointerface);
  std::set<std::string> preferredInterfaceNames;
  if (usercfg.InfoInterface().empty()) {
    preferredInterfaceNames.insert("org.nordugrid.ldapglue2");
  } else {
    preferredInterfaceNames.insert(usercfg.InfoInterface());
  }

  Arc::ExecutionTargetSorter ets(broker);

  std::list<std::string> rejectDiscoveryURLs = getRejectDiscoveryURLsFromUserConfigAndCommandLine(usercfg, opt.rejectdiscovery);
  Arc::ComputingServiceRetriever csr(usercfg, std::list<Arc::Endpoint>(), rejectDiscoveryURLs, preferredInterfaceNames);
  csr.addConsumer(ets);
  for (std::list<Arc::Endpoint>::const_iterator it = services.begin(); it != services.end(); ++it) {
    csr.addEndpoint(*it);
  }
  csr.wait();

  if (csr.empty()) {
    if (!opt.dumpdescription) {
      std::cout << Arc::IString("Test aborted because no resource returned any information") << std::endl;
    } else {
      std::cout << Arc::IString("Dumping job description aborted because no resource returned any information") << std::endl;
    }
    return 1;
  }

  if (ets.getMatchingTargets().empty()) {
    if (!opt.dumpdescription) {
      std::cout << Arc::IString("ERROR: Test aborted because no suitable resources were found for the test-job") << std::endl;
    } else {
      std::cout << Arc::IString("ERROR: Dumping job description aborted because no suitable resources were found for the test-job") << std::endl;
    }
    return 1;
  }

  if (opt.dumpdescription) {
     return dumpjobdescription(usercfg, ets, testJob);
  }
  return test(usercfg, ets, testJob, opt.jobidoutfile);
}

void printjobid(const std::string& jobid, const std::string& jobidfile) {
  if (!jobidfile.empty())
    if (!Arc::Job::WriteJobIDToFile(jobid, jobidfile))
      logger.msg(Arc::WARNING, "Cannot write jobid (%s) to file (%s)", jobid, jobidfile);
  std::cout << Arc::IString("Test submitted with jobid: %s", jobid) << std::endl;
}

int test(const Arc::UserConfig& usercfg, Arc::ExecutionTargetSorter& ets, const Arc::JobDescription& testJob, const std::string& jobidfile) {
  int retval = 0;

  std::list<std::string> jobids;
  std::list<Arc::Job> submittedJobs;
  std::map<int, std::string> notsubmitted;

  submittedJobs.push_back(Arc::Job());

  for (ets.reset(); !ets.endOfList(); ets.next()) {
    if (ets->Submit(usercfg, testJob, submittedJobs.back())) {
      printjobid(submittedJobs.back().JobID.fullstr(), jobidfile);
      break;
    }
  }
  
  if (ets.endOfList()) {
    std::cout << Arc::IString("Test failed, no more possible targets") << std::endl;
    submittedJobs.pop_back();
    retval = 1;
  }

  if (!Arc::Job::WriteJobsToFile(usercfg.JobListFile(), submittedJobs)) {
    std::cout << Arc::IString("Warning: Failed to lock job list file %s", usercfg.JobListFile())
              << std::endl;
    std::cout << Arc::IString("To recover missing jobs, run arcsync") << std::endl;
  }

  return retval;
}

int dumpjobdescription(const Arc::UserConfig& usercfg, Arc::ExecutionTargetSorter& ets, const Arc::JobDescription& testJob) {
  for (ets.reset(); !ets.endOfList(); ets.next()) {
    Arc::JobDescription preparedTestJob(testJob);
    std::string jobdesc;
    // Prepare the test jobdescription according to the choosen ExecutionTarget
    if (!preparedTestJob.Prepare(*ets)) {
      logger.msg(Arc::INFO, "Unable to prepare job description according to needs of the target resource (%s).", ets->ComputingEndpoint->URLString); 
      continue;
    }
  
    std::string jobdesclang = "nordugrid:jsdl";
    if (ets->ComputingEndpoint->InterfaceName == "org.nordugrid.gridftpjob") {
      jobdesclang = "nordugrid:xrsl";
    }
    else if (ets->ComputingEndpoint->InterfaceName == "org.glite.cream") {
      jobdesclang = "egee:jdl";
    }
    
    if (!preparedTestJob.UnParse(jobdesc, jobdesclang)) {
      logger.msg(Arc::INFO, "An error occurred during the generation of job description to be sent to %s", ets->ComputingEndpoint->URLString); 
      continue;
    }
  
    std::cout << Arc::IString("Job description to be sent to %s:", ets->AdminDomain->Name) << std::endl;
    std::cout << jobdesc << std::endl;
    break;
  }

  return (!ets.endOfList());
}
